core/opendaq/opendaq/src/instance_builder_impl.cpp

```diff
@@ -0,0 +1,303 @@
+#include <coreobjects/errors.h>
+#include <opendaq/instance_builder_impl.h>
+#include <opendaq/instance_builder_ptr.h>
+#include <opendaq/instance_ptr.h>
+#include <utility>
+
+BEGIN_NAMESPACE_OPENDAQ
+
+DictPtr<IString, IBaseObject> InstanceBuilderImpl::GetOptions()
+{
+    return Dict<IString, IBaseObject>({
+        {"ModuleManager", Dict<IString, IBaseObject>({
+                {"ModulesPath", ""}
+            })},
+        {"Scheduler", Dict<IString, IBaseObject>({
+                {"WorkersNum", 0}
+            })},
+        {"Logging", Dict<IString, IBaseObject>({
+                {"GlobalLogLevel", OPENDAQ_LOG_LEVEL_DEFAULT}
+            })},
+        {"Modules", Dict<IString, IBaseObject>()}
+    });
+}
+
+InstanceBuilderImpl::InstanceBuilderImpl()
+    : componentsLogLevel(Dict<IString, LogLevel>())
+    , options(GetOptions())
+{
+}
+
+DictPtr<IString, IBaseObject> InstanceBuilderImpl::getModuleManagerOptions()
+{
+    return options.get("ModuleManager");
+}
+DictPtr<IString, IBaseObject> InstanceBuilderImpl::getSchedulerOptions()
+{
+    return options.get("Scheduler");
+}
+DictPtr<IString, IBaseObject> InstanceBuilderImpl::getLoggingOptions()
+{
+    return options.get("Logging");
+}
+DictPtr<IString, IBaseObject> InstanceBuilderImpl::getModuleOptions(IString* module)
+{
+    DictPtr<IString, IBaseObject> modules = options.get("Modules").asPtr<IDict>();
+    if (!modules.hasKey(module)) 
+    {   
+        auto moduleOptions = Dict<IString, IBaseObject>();
+        modules[module] = moduleOptions;
+        return moduleOptions;
+    }
+    return modules[module];
+}
+
+ErrCode InstanceBuilderImpl::build(IInstance** instance)
+{
+    if (instance == nullptr)
+        return OPENDAQ_ERR_ARGUMENT_NULL;
+
+    const auto builderPtr = this->borrowPtr<InstanceBuilderPtr>();
+    return daqTry([&]()
+    {
+        *instance = InstanceFromBuilder(builderPtr).detach();
+        return OPENDAQ_SUCCESS;
+    });
+}
+
+ErrCode InstanceBuilderImpl::setLogger(ILogger* logger)
+{
+    this->logger = logger;    
+    return OPENDAQ_SUCCESS;
+}
+
+ErrCode InstanceBuilderImpl::getLogger(ILogger** logger)
+{
+    if (logger == nullptr)
+        return OPENDAQ_ERR_ARGUMENT_NULL;
+    
+    if (this->logger.assigned()) 
```

Comment:
logger.addRefAndReturn() will return nullptr if logger not assigned. No need to handle not assigned case here.

Reply:
This still applies - the `if (this->logger.assigned())` check is redundant here